// externality/dist/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const url = require('url');
const util = require('util');
const ufo = require('ufo');
const enhancedResolve = require('enhanced-resolve');
const mlly = require('mlly');
const pathe = require('pathe');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e["default"] : e; }

const enhancedResolve__default = /*#__PURE__*/_interopDefaultLegacy(enhancedResolve);

const ProtocolRegex = /^(?<proto>.{2,}):.+$/;
function getProtocol(id) {
  const proto = id.match(ProtocolRegex);
  return proto ? proto.groups.proto : null;
}
function matches(input, matchers, ctx) {
  return matchers.some((matcher) => {
    if (matcher instanceof RegExp) {
      return matcher.test(input);
    }
    if (typeof matcher === "function") {
      return matcher(input, ctx);
    }
    return false;
  });
}
function toMatcher(pattern) {
  if (typeof pattern !== "string") {
    return pattern;
  }
  pattern = pattern.replace(/\//g, "[\\\\/+]");
  return new RegExp(`([\\/]|^)${pattern}(@[^\\/]*)?([\\/](?!node_modules)|$)`);
}
function getType(id, fallback = "commonjs") {
  if (id.endsWith(".cjs")) {
    return "commonjs";
  }
  if (id.endsWith(".mjs")) {
    return "module";
  }
  return fallback;
}

const DefaultResolveOptions = {
  extensions: [".ts", ".mjs", ".cjs", ".js", ".json"],
  type: "commonjs"
};
async function resolveId(id, base = ".", opts = {}) {
  opts = { ...DefaultResolveOptions, ...opts };
  if (!opts.conditionNames) {
    opts.conditionNames = [opts.type === "commonjs" ? "require" : "import"];
  }
  if (mlly.isNodeBuiltin(id)) {
    return {
      id: id.replace(/^node:/, ""),
      path: id,
      type: opts.type,
      external: true
    };
  }
  if (id.startsWith("file:/")) {
    id = url.fileURLToPath(id);
  }
  if (ufo.hasProtocol(id)) {
    const url = new URL(id);
    return {
      id: url.href,
      path: url.pathname,
      type: getType(id, opts.type),
      external: true
    };
  }
  if (base.includes("\0")) {
    base = opts.roots?.[0] || ".";
  }
  const _resolve = util.promisify(enhancedResolve__default.create(opts));
  const resolvedModule = await _resolve(base, id);
  return {
    id,
    path: resolvedModule,
    type: getType(resolvedModule, "unknown")
  };
}

const ExternalsDefaults = {
  inline: [
    /\x00/,
    /^!/,
    /^-!/,
    /^#/,
    /\?/
  ],
  external: [],
  externalProtocols: ["node", "file", "data"],
  externalExtensions: [".js", ".mjs", ".cjs", ".node"],
  resolve: {},
  detectInvalidNodeImports: true
};
async function isExternal(id, importer, opts = {}) {
  opts = { ...ExternalsDefaults, ...opts };
  const inlineMatchers = opts.inline.map(toMatcher);
  const externalMatchers = opts.external.map(toMatcher);
  const ctx = { opts, id, resolved: null };
  if (!id || matches(id, inlineMatchers, ctx)) {
    return null;
  }
  const proto = getProtocol(id);
  if (proto && !opts.externalProtocols.includes(proto)) {
    return null;
  }
  if (proto === "data") {
    return { id, external: true };
  }
  const r = ctx.resolved = await resolveId(id, importer, opts.resolve).catch((_err) => {
    return { id, path: id, external: null };
  });
  const idExt = pathe.extname(r.path);
  if (idExt && !opts.externalExtensions.includes(idExt)) {
    return null;
  }
  if (matches(r.id, inlineMatchers, ctx) || matches(r.path, inlineMatchers, ctx)) {
    return null;
  }
  if (r.external || matches(id, externalMatchers, ctx) || matches(r.id, externalMatchers, ctx) || matches(r.path, externalMatchers, ctx)) {
    if (opts.detectInvalidNodeImports && !await mlly.isValidNodeImport(r.path)) {
      return null;
    }
    return { id: r.id, external: true };
  }
  return null;
}

function rollupExternals(opts) {
  return {
    name: "node-externals",
    resolveId(id, importer) {
      return isExternal(id, importer, opts);
    }
  };
}

function webpackExternals(opts) {
  const _isExternal = async ({ request }, cb) => {
    try {
      const res = await isExternal(request, ".", opts);
      cb(void 0, res && res.id);
    } catch (err) {
      cb(err, null);
    }
  };
  return _isExternal;
}

exports.ExternalsDefaults = ExternalsDefaults;
exports.getProtocol = getProtocol;
exports.getType = getType;
exports.isExternal = isExternal;
exports.matches = matches;
exports.resolveId = resolveId;
exports.rollupExternals = rollupExternals;
exports.toMatcher = toMatcher;
exports.webpackExternals = webpackExternals;
