// unimport/dist/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const context = require('./chunks/context.cjs');
require('fs');
require('globby');
require('pathe');
require('mlly');
require('scule');
require('magic-string');
require('local-pkg');



exports.addImportToCode = context.addImportToCode;
exports.builtinPresets = context.builtinPresets;
exports.createUnimport = context.createUnimport;
exports.dedupeImports = context.dedupeImports;
exports.defineUnimportPreset = context.defineUnimportPreset;
exports.excludeRE = context.excludeRE;
exports.getMagicString = context.getMagicString;
exports.getString = context.getString;
exports.importAsRE = context.importAsRE;
exports.matchRE = context.matchRE;
exports.normalizeImports = context.normalizeImports;
exports.resolveBuiltinPresets = context.resolveBuiltinPresets;
exports.resolveFiles = context.resolveFiles;
exports.resolvePreset = context.resolvePreset;
exports.scanDirExports = context.scanDirExports;
exports.scanExports = context.scanExports;
exports.separatorRE = context.separatorRE;
exports.stripCommentsAndStrings = context.stripCommentsAndStrings;
exports.toExports = context.toExports;
exports.toImports = context.toImports;
exports.toTypeDeclrationFile = context.toTypeDeclrationFile;
exports.toTypeDeclrationItems = context.toTypeDeclrationItems;
