// unenv/runtime/npm/whatwg-url.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setTheUsername = exports.setThePassword = exports.serializeURLOrigin = exports.serializeURL = exports.serializeInteger = exports.serializeHost = exports.percentDecodeString = exports.percentDecodeBytes = exports.parseURL = exports.cannotHaveAUsernamePasswordPort = exports.basicURLParse = exports.URLSearchParams = exports.URL = void 0;

var _utils = require("../_internal/utils");

const URL = globalThis.URL;
exports.URL = URL;
const URLSearchParams = globalThis.URLSearchParams;
exports.URLSearchParams = URLSearchParams;
const parseURL = (0, _utils.notImplemented)("whatwg-url.parseURL");
exports.parseURL = parseURL;
const basicURLParse = (0, _utils.notImplemented)("whatwg-url.basicURLParse");
exports.basicURLParse = basicURLParse;
const serializeURL = (0, _utils.notImplemented)("whatwg-url.serializeURL");
exports.serializeURL = serializeURL;
const serializeHost = (0, _utils.notImplemented)("whatwg-url.serializeHost");
exports.serializeHost = serializeHost;
const serializeInteger = (0, _utils.notImplemented)("whatwg-url.serializeInteger");
exports.serializeInteger = serializeInteger;
const serializeURLOrigin = (0, _utils.notImplemented)("whatwg-url.serializeURLOrigin");
exports.serializeURLOrigin = serializeURLOrigin;
const setTheUsername = (0, _utils.notImplemented)("whatwg-url.setTheUsername");
exports.setTheUsername = setTheUsername;
const setThePassword = (0, _utils.notImplemented)("whatwg-url.setThePassword");
exports.setThePassword = setThePassword;
const cannotHaveAUsernamePasswordPort = (0, _utils.notImplemented)("whatwg-url.cannotHaveAUsernamePasswordPort");
exports.cannotHaveAUsernamePasswordPort = cannotHaveAUsernamePasswordPort;
const percentDecodeBytes = (0, _utils.notImplemented)("whatwg-url.percentDecodeBytes");
exports.percentDecodeBytes = percentDecodeBytes;
const percentDecodeString = (0, _utils.notImplemented)("whatwg-url.percentDecodeString");
exports.percentDecodeString = percentDecodeString;
